Add retry mechanism and dataframe support for WeaviateIngestOperator

utkarsharma2 opened the pull request:
The PR add the retry mechanism while ingesting data into Weaviate, this is needed to ensure the data that we intend to ingest is not affected by the possible network or ratelimiting issue. We also extend the support for ingestion data format to include pandas dataframe.

utkarsharma2 committed:
Add retry and dataframe support

uranusjr commented on airflow/providers/weaviate/hooks/weaviate.py:
```diff
@@ -135,22 +139,40 @@ def create_schema(self, schema_json: dict[str, Any]) -> None:
         client.schema.create(schema_json)
 
+    @staticmethod
+    def check_http_error_is_retryable(exc: BaseException):
```
```suggestion
    def check_http_error_should_retry(exc: BaseException):
```

The `not exc.response.ok` part makes the “retryable” part not accurate (technically an OK response is still retryable?)

uranusjr commented on airflow/providers/weaviate/hooks/weaviate.py:
```diff
@@ -135,22 +139,40 @@ def create_schema(self, schema_json: dict[str, Any]) -> None:
-                    batch.add_data_object(data_obj, class_name)
+                for attempt in Retrying(
+                    stop=stop_after_attempt(no_retry_attempts_per_object),
+                    retry=retry_if_exception(lambda exc: self.check_http_error_is_retryable(exc)),
```
```suggestion
                    retry=retry_if_exception(self.check_http_error_is_retryable),
```

uranusjr commented on docs/apache-airflow-providers-weaviate/operators/weaviate.rst:
```diff
@@ -28,7 +28,7 @@ into the database.
 ^^^^^^^^^^^^^^^^^^
 
-The WeaviateIngestOperator requires the ``input_text`` as an input to the operator. Use the ``conn_id`` parameter to specify the Weaviate connection to use to
+The WeaviateIngestOperator requires the ``input_data`` as an input to the operator. Use the ``conn_id`` parameter to specify the Weaviate connection to use to
```
Previously the operator had `input_json` and now `input_data`. Does this count as breaking compatibility? And the doc had `input_text` and now `input_data`. Was the old content simply wrong? It may be worthwhile to dig out some history on why the doc and code did not match.

utkarsharma2 replied:
Yes, that's true. I'll add a deprecation warning for `input_json` and introduce `input_data` while supporting both parameters. Thanks for pointing this out.

Also, I'll check on the history of docs.

utkarsharma2 replied:
I'm afraid the old content in the docs was wrong, input param `input_text` was never used. I guess it was missed while addressing PR comments, we updated the code but not the doc.

utkarsharma2 committed:
Update airflow/providers/weaviate/hooks/weaviate.py

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

utkarsharma2 committed:
Update airflow/providers/weaviate/hooks/weaviate.py

Co-authored-by: Tzu-ping Chung <uranusjr@gmail.com>

utkarsharma2 commented on airflow/providers/weaviate/hooks/weaviate.py:
```diff
@@ -135,22 +139,40 @@ def create_schema(self, schema_json: dict[str, Any]) -> None:
-                    batch.add_data_object(data_obj, class_name)
+                for attempt in Retrying(
+                    stop=stop_after_attempt(no_retry_attempts_per_object),
+                    retry=retry_if_exception(self.check_http_error_is_retryable),
```
```suggestion
                    retry=retry_if_exception(check_http_error_should_retry),
```

utkarsharma2 committed:
Update airflow/providers/weaviate/hooks/weaviate.py

utkarsharma2 committed:
Add deprication warning and fix tests

utkarsharma2 committed:
Add 'input_json' param in docstring

pankajkoti commented on airflow/providers/weaviate/hooks/weaviate.py:
```diff
@@ -135,22 +139,40 @@ def create_schema(self, schema_json: dict[str, Any]) -> None:
+        data: list[dict[str, Any]] | pd.DataFrame,
+        batch_config_params: dict[str, Any] | None = None,
+        vector_col: str = "Vector",
+        no_retry_attempts_per_object: int = 5,
```
```suggestion
        retry_attempts_per_object: int = 5,
```
I am guessing you meant no=number, but looking at the first time I read it as NO (some negation).

utkarsharma2 replied:
yes, no=number. Added your suggestion.

pankajkoti commented on airflow/providers/weaviate/operators/weaviate.py:
```diff
@@ -40,9 +44,11 @@ class WeaviateIngestOperator(BaseOperator):
     :param class_name: The Weaviate class to be used for storing the data objects into.
+    :param input_data: The list of dicts or pandas dataframe representing Weaviate data objects to generate
+        embeddings on (or provides custom vectors) and store them in the Weaviate class.
     :param input_json: The JSON representing Weaviate data objects to generate embeddings on (or provides
```
Can we update the docstring to mark this as deprecated?

utkarsharma2 replied:
Oops missed it. Now added `(Deprecated)` at the start of the param string.

pankajkoti commented on airflow/providers/weaviate/hooks/weaviate.py:
```diff
@@ -135,22 +139,40 @@ def create_schema(self, schema_json: dict[str, Any]) -> None:
+        class_name: str,
+        data: list[dict[str, Any]] | pd.DataFrame,
+        batch_config_params: dict[str, Any] | None = None,
+        vector_col: str = "Vector",
```
I think it would make sense to add docstring for this public method of the hook where we explain what these params mean.

utkarsharma2 replied:
Thanks for pointing that out, I added those.

utkarsharma2 committed:
Fix docstrings

utkarsharma2 committed:
Fix static checks

utkarsharma2 committed:
Add the type: ignore at right line

utkarsharma2 commented on airflow/providers/weaviate/operators/weaviate.py:
```diff
@@ -51,21 +57,38 @@ def __init__(
+        self.log.debug("Input data: %s", self.input_data)
+        self.hook.batch_data(
+            self.class_name,
+            self.input_data,  # type: ignore
```
@uranusjr I added `#  type: ignore` because mypy was assuming the type of `self.input_data` to be `list[dict[str, Any]] | pd.DataFrame | None` but we have ensured in the `__init__()` method that it's not None. Is there a better way of dealing with it?

utkarsharma2 replied:
cc: @ephraimbuddy @pankajastro ?

uranusjr replied:
I tweaked a check to make Mypy understand the attribute can never be None. The trick is to move the check before assignment. Since Mypy infers attribute types on assignment time, if you check after assignment, Mypy will still assume the None case is needed because technically you can do e.g.

```python
op = WeaviateIngestOperator(input_data=[...])
op.input_data = None  # Takes advantage of the None case.
op.execute({})
```

so we need to eliminate the None case before the assignment so Mypy does not consider the None case to be valid on the attribute at all.

utkarsharma2 replied:
That's a clever solution, thanks! :)

Also, you changed the exception to `TypeError` from `ValueError`, was there any reason for that? To my understanding, we should raise a `ValueError` right?

utkarsharma2 committed:
Fix typing

uranusjr committed:
Tweak checks in __init__ to hint on None-ness

By moving the check to before the self.input_data assignment, we can
hint to the checker that this attribute is never None.

uranusjr commented on airflow/providers/weaviate/hooks/weaviate.py:
```diff
@@ -135,22 +141,52 @@ def create_schema(self, schema_json: dict[str, Any]) -> None:
+        :param vector_col: name of the column containing the vector.
+        :param retry_attempts_per_object: number of time to try in case of failure before giving up.
+        """
+        import pandas as pd
```
Is it possible for Pandas to _not_ be installed? Because this might fail in this case.

utkarsharma2 replied:
Yes, it will fail if Pandas is not installed. I saw the same usage across providers. But I guess we should add pandas in the dependencies section of provider.yaml?

uranusjr replied:
Since Weaviate does not strictly require Pandas to function, it would be better to do something like

```python
with contextlib.suppress(ImportError):
    import pandas

    if isinstance(data, pandas.DataFrame):
        ...
```

If the import fails, `data` can never be a DataFrame (it’s impossible to create without Pandas installed), so we can safely guard the check in a try-except.

utkarsharma2 replied:
Sure, makes sense. I moved this logic out into a different method, as it will be used in other methods. 

Also, we probably don't need a try except for a check, right?

utkarsharma2 committed:
Fix typing issue

utkarsharma2 committed:
Handle case when pandas is not installed

utkarsharma2 committed:
Fix typing issue

ephraimbuddy commented on airflow/providers/weaviate/hooks/weaviate.py:
```diff
@@ -17,10 +17,14 @@
 import warnings
 from functools import cached_property
-from typing import TYPE_CHECKING, Sequence
+from typing import TYPE_CHECKING, Any, Dict, List, cast
```
```suggestion
from typing import TYPE_CHECKING
```

utkarsharma2 replied:
we need this outside of `if TYPE_CHECKING` because are using them in `cast(List[Dict[str, Any]], data)` statement.

ephraimbuddy commented on airflow/providers/weaviate/hooks/weaviate.py:
```diff
@@ -30,7 +34,7 @@
 
 if TYPE_CHECKING:
-    from typing import Any
+    from typing import Sequence
```
```suggestion
    from typing import Sequence, Any, Dict, List, cast
```

utkarsharma2 replied:
https://github.com/apache/airflow/pull/36085#discussion_r1420266560

ephraimbuddy commented on airflow/providers/weaviate/hooks/weaviate.py:
```diff
@@ -144,22 +148,63 @@ def create_schema(self, schema_json: dict[str, Any]) -> None:
+                        if vector is not None:
+                            batch.add_data_object(data_obj, class_name, vector=vector)
+                        else:
+                            batch.add_data_object(data_obj, class_name)
```
```suggestion
                        batch.add_data_object(data_obj, class_name, vector=vector)
```
There's no need for the if statement because, by default, the vector param is None, so we're good. I'm just worried if we need to do a `pop` instead of `get` for the `vector`?

utkarsharma2 replied:
We are already doing `pop` I guess we can just remove the `if-else` block.

utkarsharma2 committed:
Fix static issue

utkarsharma2 committed:
Remove unwanted if-else block

utkarsharma2 committed:
Merge branch 'main' into retry_batch_data
